Sources/Views/MessageLabel.swift

```diff
@@ -22,417 +22,491 @@
+    
+    private func updateDetectorAttributes(for key: NSAttributedStringKey? = nil) {
+        
+        guard !enabledDetectors.isEmpty else { return }
+        
+        let selectedAttributes = { (attributes: [NSAttributedStringKey: Any], key: NSAttributedStringKey?)
+                            -> [NSAttributedStringKey: Any]? in
+            
+            // Key was found so we are applying selective attributes
+            if let key = key {
+                // No attributes exist for this key, exit early
+                guard let value = attributes[key] else { return nil }
+                return [key: value]
+            } else {
+                return attributes
+            }
+        
         }
 
+        for detector in enabledDetectors {
+            if let ranges = detectedRanges(for: detector) {
+                let attributes = detectorAttributes(for: detector)
+                guard let selected = selectedAttributes(attributes, key) else { continue }
+                applyAttributes(selected, to: ranges)
+            }
+        }
+        
     }
+    
+    private func setStorageByParsing(text: NSAttributedString?, completion: @escaping () -> Void) {
+        
+        removeAllDetectedResults()
+
+        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
+            
+            defer {
+                DispatchQueue.main.async {
+                    completion()
+                }
+            }
+            
+            guard let `self` = self else { return }
+            guard let text = text, text.length > 0, !self.enabledDetectors.isEmpty else { return }
+            guard let checkingResults = self.parse(text: text.string) else { return }
+  
+            self.setDetectorResults(for: checkingResults)
+            self.updateDetectorAttributes()
 
-    private func detectorAttributes(for checkingResultType: NSTextCheckingResult.CheckingType) -> [NSAttributedStringKey: Any] {
-        switch checkingResultType {
-        case NSTextCheckingResult.CheckingType.address:
-            return addressAttributes
-        case NSTextCheckingResult.CheckingType.date:
-            return dateAttributes
-        case NSTextCheckingResult.CheckingType.phoneNumber:
-            return phoneNumberAttributes
-        case NSTextCheckingResult.CheckingType.link:
-            return urlAttributes
-        default:
-            fatalError("Received an unrecognized NSTextCheckingResult.CheckingType")
         }
+        
     }
-
-    // MARK: - Parsing Text
-
-    private func parse(text: NSAttributedString, for detectorTypes: [DetectorType]) -> [NSTextCheckingResult]? {
-        guard detectorTypes.isEmpty == false else { return nil }
-        let checkingTypes = detectorTypes.reduce(0) { $0 | $1.textCheckingType.rawValue }
+    
+    private func parse(text: String) -> [NSTextCheckingResult]? {
+        let checkingTypes = enabledDetectors.reduce(0) { $0 | $1.textCheckingType.rawValue }
```

Comment:
I would add `if enabledDetectors.isEmpty { return nil }` here as well

Reply:
Yeah, I'm struggling with this. Some of these methods are called within over methods where the check has already been performed. I started removing them because of this